### PR TITLE
8296139: Make GrowableBitMap the base class of all implementations

### PR DESCRIPTION
**Pull request by navyxliu**
BitMap is an abstract class. All subclasses provide their own allocators. It is the allocator brings the capability to 'resize' the container. 
I would like to add a new 'GrowableBitmap' and make it the common base class. 

To substitute 'VectorSet', we also need to merge ResourceBitMap and ArenaBitMap because VectorSet supports Resource and Arena storage at the same time. 
This patch unifies them.  ResourceBitMap is a specialized ArenaBitMap whose allocator is with NULL arena.

All subclasses of GrowableBitMap are elastic. I will add new interface like VectorSet::test_set(index). It will grow to include the new index.
/cc hotspot-runtime

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296139](https://bugs.openjdk.org/browse/JDK-8296139): Make GrowableBitMap the base class of all implementations

### Reviewers
 * [Axel Boldt-Christmas](https://openjdk.org/census#aboldtch) (@xmas92 - Committer)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Volker Simonis](https://openjdk.org/census#simonis) (@simonis - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10941/head:pull/10941` \
`$ git checkout pull/10941`

Update a local copy of the PR: \
`$ git checkout pull/10941` \
`$ git pull https://git.openjdk.org/jdk pull/10941/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10941`

View PR using the GUI difftool: \
`$ git pr show -t 10941`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10941.diff">https://git.openjdk.org/jdk/pull/10941.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/bitMap.hpp
Comment (xmas92) [inline src/hotspot/share/utilities/bitMap.hpp:412]: How do we want to treat resource allocations which look like arena allocations? The compiler code does have a places where objects are arena allocated in `Thread::current()->resource_area()`. I am not sure about having this behaviour in the type system. 

The places where `VectorSet()` is used today, can we not just use `ArenaBitMap(Thread::current()->resource_area(), [...])` instead. And have the replacement be `VectorSet` -> `ArenaBitMap`

**Review by xmas92 — COMMENTED** [review]
I am curious how the final type hierarchy will look here. 

I am guessing the `GrowableBitMap` will get something like `set_bit_grow` etc. 
Given that more than just `resize` and `[re]initialize` will now change the underlying storage maybe we should have the same/similar abstraction as `GrowableArray`. 
At least the separating the code that does not change the size and is allocator agnostic (like `GrowableArrayView`) and the code which may grow and deallocate (like `GrowableArrayWithAllocator`). 

Other than questions of where this is going and the resource/arena thing this looks good to me, and I like the separation of the allocation logic and the pure bitmap logic.

P.S. Maybe someone more knowledgable than I can comment on this style of using allocators vs the CRTP style used in `GrowableArray`. The CRTP seems more encapsulating to me.

File: src/hotspot/share/utilities/bitMap.hpp
Comment (navyxliu) [reply src/hotspot/share/utilities/bitMap.hpp:412]: This is how I treat them using the base pointer 'ArenaBitMap*'
https://github.com/navyxliu/jdk/commit/cdbc0e1c97f747fd53ad408d59e65ec8adf0fc16#diff-17a9d48a5570e5bcde557c8d1efcbe78d69a5657bc865bed37b22c3c8b5d447d

> The places where VectorSet() is used today, can we not just use ArenaBitMap(Thread::current()->resource_area(), [...]) instead. And have the replacement be VectorSet -> ArenaBitMap

`Thread::current()->resource_area()` returns ResourceArena*. ResourceArena is a subclass of Arena. In this sense, we can use the same ArenaAllocator to handle both ResourceBitMap and ArenaBitMap. That's why I remove 'ResourceAllocator' in this patch. 

I think it is possible to do what you said. Let's say we use ArenaBitMap as 'ResourceBitMap'. Shall we keep ArenaBitMap and ResourceBitMap separated then? In classic object-oriented design, ResourceBitMap 'is-a' ArenaBitMap, isn't it? If we keep 2 classes in parallel, we may end up 2 'ResourceBitMap' with different behaviors. 

In current patch, ResourceBitMap is just an adapter. It supports copy constructor and copy assignment because 'resource arena' is unique.

Comment (navyxliu) [reply src/hotspot/share/utilities/bitMap.hpp:412]: Correction: s/resource arena/resource area

This is my understanding. 'Resource area' in HotSpot is a special 'area' whose lifecycle is bound to stacktrace. By default, we still use 'arena' allocation algorithm. That's why I think we can treat ResourceBitMap as a subclass of ArenaBitMap.

Comment (xmas92) [reply src/hotspot/share/utilities/bitMap.hpp:412]: I like the type hierarchy but for one thing. Today we have two ways of creating a resource arena
* `ResourceBitmap([...])`
  * Which checks `verify_has_resource_mark` in debug builds
* `ArenaBitmap(some_resource_area, [...])`
  * Which does not check `verify_has_resource_mark`

This would add a third way where `ArenaBitmap(nullptr, [...])` has the same behaviour as `ResourceBitmap`. This can be solved with a new `ArenaBitMap` constructor.
```c++
 protected:
  ArenaBitMap(std::nullptr_t, idx_t size_in_bits, bool clean);
```
And add a debug assert in the normal ArenaBitMap constructor `precond(arena != nullptr);`

Which would give a behaviour like this
```c++
#ifdef ASSERT
TEST_VM(BitMap, resource_work) {
  ResourceBitMap rbm(0);
  ArenaBitMap abm(Thread::current()->resource_area(),0);
  static_cast<void>(rbm);
  static_cast<void>(abm);
}
TEST_VM_ASSERT(BitMap, nullptr_arena) {
  Arena* ptr = nullptr;
  ArenaBitMap crash(ptr,0);
  // ArenaBitMap does_not_compile(nullptr,0);
  // ArenaBitMap does_not_compile(NULL,0);
  static_cast<void>(crash);
}
#endif
```

I do not know if the extra branch in the arena allocator ever matters, but if it did you can just store the resource area arena in `_arena` on release and not branch. The only semantical difference then would be that if the [re]allocating functions gets called from different threads. But I would think that a resource allocated bitmap crossing thread boundaries is never an appropriate use. 

But I did preface this discussion with that I do not know how we treat resource allocated objects that just look like arena allocated objects and thus circumvent the `verify_has_resource_mark` check. It happens all over the place in the compiler code, and I am not sure why.

File: src/hotspot/share/utilities/bitMap.hpp
Comment (stefank) [inline src/hotspot/share/utilities/bitMap.hpp:330]: GrowableBitMap(idx_t, bool) is dangerous to use because it calls reallocate, which calls the allocate function of the subclass. This works for ResourceBitMap, but not if you tried to use it for CHeapBitMap or ArenaBitMap. I'd suggest that we never let GrowableBitMap allocate memory in the constructor and instead push that responsiblity down into the subclasses.

File: src/hotspot/share/utilities/bitMap.cpp
Comment (stefank) [inline src/hotspot/share/utilities/bitMap.cpp:91]: Both CHeapBitMap and AreanBitMap perform this "initialization". I propose that this instead call the GrowableBitMap::initialize function.

**Review by stefank — CHANGES_REQUESTED** [review]
I started to review this PR and added comments, but I think it will be easier if you take a look at this branch which contains all my proposed changes + some small cleanups:
https://github.com/stefank/jdk/tree/pr_10941

File: src/hotspot/share/utilities/bitMap.cpp
Comment (navyxliu) [reply src/hotspot/share/utilities/bitMap.cpp:91]: I agree! 'initialize()' is clearer to reflect the purpose of code.

File: src/hotspot/share/utilities/bitMap.hpp
Comment (navyxliu) [reply src/hotspot/share/utilities/bitMap.hpp:170]: Confirm that this got lost in shuffle. `verify_size` is zero-cost in release build. I will put it back. Thanks. 

I also notice that some static member functions such as `raw_to_words_align_down` can be decorated using constexpr. We could change them in a separated JBS.

File: src/hotspot/share/utilities/bitMap.hpp
Comment (navyxliu) [reply src/hotspot/share/utilities/bitMap.hpp:330]: agree. I see your patch has changed this. This is safer. 
```
  GrowableBitMap() : GrowableBitMap(nullptr, 0) {}
  GrowableBitMap(bm_word_t* map, idx_t size_in_bits) : BitMap(map, size_in_bits) {}
```

**Comment by stefank** [comment]
Looks good to me. Make sure to get one more review for this patch.

**Review by stefank — APPROVED** [review]

File: src/hotspot/share/utilities/bitMap.hpp
Comment (navyxliu) [reply src/hotspot/share/utilities/bitMap.hpp]: First of all, I revert back ResourceBitMap. It's now independent to ArenaBitMap. I think you concern is reasonable. In most cases, I expect users to use ResourceBitMap. ArenaBitmap does accept ResourceArea* as the underlying arena because Unique_Node_List uses it in this way. 

I don't think we need to handle ArenaBitMap::_arena == nullptr like you said. ResourceArea is still an Arena, but is managed by ResourceMark automatically. As a result, I introduce a boolean flag `Arena::_mark_managed`. ArenaBitMap doesn't mess with the underlying storage if Arena::mark_managed() is true. 

Here is my experimental patch.
https://github.com/navyxliu/jdk/commit/88063064a3bc0b8e65d3834f13e5ea8c6576eab0#diff-17a9d48a5570e5bcde557c8d1efcbe78d69a5657bc865bed37b22c3c8b5d447d

We have to support copy assignment for Unique_Node_List. It won't introduce memory leak because we only use ArenaBitMap in certain scenarios, see the assert. Besides ResourceArena, Unique_Node_List only accepts Compile::comp_arena() || Compiler::node_arena(). The lifecycle of them are same as Compile object. once a compilation unit is done, both two arenas are released completely.

File: src/hotspot/share/utilities/bitMap.hpp
Comment (xmas92) [inline src/hotspot/share/utilities/bitMap.hpp:401]: s/ArenaBitMaps/ResourceBitMap/

Or keep the old `Don't free resource allocated arrays.` comment

**Review by xmas92 — APPROVED** [review]
Lgtm. 

Just one comment, on one comment.

File: src/hotspot/share/utilities/bitMap.hpp
Comment (navyxliu) [reply src/hotspot/share/utilities/bitMap.hpp:401]: I take a note and will fix this one in the next change. thanks!

File: src/hotspot/share/utilities/bitMap.cpp
Comment (simonis) [inline src/hotspot/share/utilities/bitMap.cpp:637]: Do you really need this explicit template instantiations? I think such code is only required in libraries where you want to be sure to have specific instantiations even though they are not used within the library itself. But here you will get these instantiations anyway whenever `ArenaBitMap`, `ResourceBitMap` or `CHeapBitMap` is used because they are derived from `GrowableBitMap<ArenaBitMap>`, `GrowableBitMap<ResourceBitMap>` and `GrowableBitMap<CHeapBitMap>` respectively.

**Review by simonis — CHANGES_REQUESTED** [review]
In general looks good. I have a single, small comment. Once we resolve that, I'll approve the change.

File: src/hotspot/share/utilities/bitMap.cpp
Comment (navyxliu) [reply src/hotspot/share/utilities/bitMap.cpp:637]: hi, Volker,

Comment (simonis) [reply src/hotspot/share/utilities/bitMap.cpp:637]: I'll take my comment back :)

It turns out that we need these explicit instantiations, because implicit instantiations do not instantiate unused members (i.e. "*This applies to the members of the class template: unless the member is used in the program, it is not instantiated, and does not require a definition*", see https://en.cppreference.com/w/cpp/language/class_template#Implicit_instantiation).

E.g.:
```
CHeapBitMap::CHeapBitMap(idx_t size_in_bits, MEMFLAGS flags, bool clear)
  : GrowableBitMap<CHeapBitMap>(), _flags(flags) {
  initialize(size_in_bits, clear);
}
```
implicitely instantiates `GrowableBitMap<CHeapBitMap>` but it will not instantiate `GrowableBitMap<CHeapBitMap>::reallocate()` because it is not used in that compilation unit.

**Review by simonis — APPROVED** [review]

**Pull request closed without merging**
